NitroxServer/Communication/UdpServer.cs

```diff
@@ -156,8 +156,30 @@ private Connection GetConnection(long remoteIdentifier)
         private NetPeerConfiguration BuildNetworkConfig()
         {
             NetPeerConfiguration config = new NetPeerConfiguration("Nitrox");
-            config.Port = 11000;
-            config.MaximumConnections = 100;
+            string INIContents;
+            int PortNumber, MaxConn;
+            if (File.Exists(@".\config.ini"))
```

Comment:
I created a new class to read the config, but do I call it before server.Start(), or call it when config.Port and config.MaximumConnections are being set

Reply:
Ideally, we should pass a custom config class into the UdpServer constructor so that file reading happens before the `UdpServer` is created.

Something like: `new UdpServer(..., UdpServerConfig.FromFile("config.ini"));` and then in the `FromFile` function you use the IniReader/Writer class you created and store the relevant INI fields in the custom config class.

Lastly, map the custom config class to `NetPeerConfiguration` inside UdpServer constructor so that Lidgren classes (like `NetPeerConfiguration`) don't leave the `UdpServer` class.

To answer your question: [`Server.Start()`](https://github.com/SubnauticaNitrox/Nitrox/blob/4cececc388920cef22b0f83313b25ff01ee0a939/NitroxServer/Server.cs#L44) sounds like the best location because reading files is slow. Move the `new UdpServer()` from the `Server` class constructor into the `Server.Start()` to do this.

Reply:
I'm not sure if the config class should be specific to UdpServer.  I think that we will have a wide range of server configurations that we'll want to access in the future. (An example being how often we auto save).  When the server first starts we can parse the config file and pass around a data structure that encapsulates all server configs.  A subset of these configs could be NetworkConfigs.

Reply:
You're right. A more abstract server config file would be better. But I'm not sure how much time  @FahadAhmad2001 has to implement all this. Please let us know what you think.

Reply:
I tried moving the new UdpServer() to server.Start(), but then I have to change it from private readonly UdpServer to private UdpServer. Also, I cant make Server.Start() static because then EnablePeriodicSaving() does not work (unless I make that static as well).

So, should I read from the config file before UdpServer is created, ie in Server.cs
<img src="https://i.gyazo.com/9fc638d7d17ec6bc23486adda0b06273.png">
This way, values such as time before auto save can also be read and set here
The code in UdpServer.cs:
<img src="https://i.gyazo.com/9e4b28dd4970ec918ea02d4f701247fd.png">
and
<img src="https://i.gyazo.com/f89a756bf2ff1a9b67eadcd88e5cbe24.png">
Also, here is the ServerConfigReader class:
<img src="https://i.gyazo.com/6422134846cfb8c807edffca96d410c7.png">
Finally, should I place the serverConfigreader class in NitroxModel or in NitroxServer?

Reply:
Looks good so far. Yes, the config should be read before the `UdpServer` and `Server` are started so that both can be configurated by the same INI file. It would be better to read the config file and pass the config class instance to the `Server` constructor instead of the file path itself. This way you don't have to bother making things static because the IO reading happens before the server constructor. I'll let you decide whether to keep `ServerConfigReader` static or pass it as an instance to the `Server` and `UdpServer` constructors. Both work.

So in short:
- Move the ServerConfigReader.ReadServerConfig(configPath) to `Program` class where the `Server` instance is created.
- Then read out the `ServerConfigReader` class in `Server` (if you have any server options in the INI file that is) and `UdpServer` (get the port from INI file).

Discard what I said about making `UdpServer.Start()` static in the comments above.